streams/src/main/java/org/apache/kafka/streams/state/internals/ReadonlyPartiallyDeserializedSegmentValue.java

```diff
@@ -0,0 +1,210 @@
+
+        // this segment does not have any record in query specified time range
+        if (toTime < minTimestamp || fromTime > nextTimestamp) {
+            return null;
+        }
+
+        final boolean isAscending = order.equals(ResultOrder.ASCENDING);
+
+        if (isAscending && valuesStartingIndex == -1) {
+            findValuesStartingIndex();
+            deserIndex = recordNumber;
+        }
+
+        long currTimestamp = -1;
+        long currNextTimestamp = -1;
+        int currIndex = initializeCurrentIndex(index, isAscending);
+        int cumValueSize = initializeCumValueSize(index, currIndex, isAscending);
+        int currValueSize;
+
+
+        while (hasStillRecord(currTimestamp, currNextTimestamp, order)) {
+            if (hasBeenDeserialized(isAscending, currIndex)) {
+                final TimestampAndValueSize curr;
+                curr = unpackedTimestampAndValueSizes.get(currIndex);
+                currTimestamp = curr.timestamp;
+                cumValueSize = cumulativeValueSizes.get(currIndex);
+                currValueSize = curr.valueSize;
+
+                // update currValueSize
+                if (currValueSize == Integer.MIN_VALUE) {
+                    final int timestampSegmentIndex = getTimestampIndex(order, currIndex);
+                    currValueSize = ByteBuffer.wrap(segmentValue).getInt(timestampSegmentIndex + TIMESTAMP_SIZE);
+                    unpackedTimestampAndValueSizes.put(currIndex, new TimestampAndValueSize(currTimestamp, cumValueSize));
+                }
+
+                currNextTimestamp = updateCurrNextTimestamp(currIndex, isAscending);
+
+            } else {
+                final int timestampSegmentIndex = getTimestampIndex(order, currIndex);
+                currTimestamp = ByteBuffer.wrap(segmentValue).getLong(timestampSegmentIndex);
+                currValueSize = ByteBuffer.wrap(segmentValue).getInt(timestampSegmentIndex + TIMESTAMP_SIZE);
+                currNextTimestamp = timestampSegmentIndex == 2 * TIMESTAMP_SIZE
+                        ? nextTimestamp // if this is the first record metadata (timestamp + value size)
+                        : ByteBuffer.wrap(segmentValue).getLong(timestampSegmentIndex - (TIMESTAMP_SIZE + VALUE_SIZE));
+                cumValueSize += Math.max(currValueSize, 0);
+
+                // update deserHelpers
+                deserIndex = currIndex;
+                unpackedTimestampAndValueSizes.put(currIndex, new TimestampAndValueSize(currTimestamp, currValueSize));
+                cumulativeValueSizes.put(currIndex, cumValueSize);
+            }
+
+            if (currValueSize >= 0) {
+                final byte[] value = new byte[currValueSize];
+                final int valueSegmentIndex = getValueSegmentIndex(order, cumValueSize, currValueSize);
+                System.arraycopy(segmentValue, valueSegmentIndex, value, 0, currValueSize);
```

Comment:
Can we move this line inside the `if` block (or maybe even all three lines?), to only do the array copy if we return a result? Otherwise, we might do this operation wastefully?